### PR TITLE
Add scripts to build python wheels

### PR DESCRIPTION
**Pull request by nafur**
This adds a new set of scripts to build python wheels for cvc5. It includes
- `mk_wheel.py` which builds the python wheel within an existing build folder
- `mk_build_dir.py` which runs `configure.sh` and directs cmake to a custom python interpreter
- `mk_clean_wheel.sh` which builds the python wheel in a clean environment (a python venv and a fresh build folder)
- a dockerfile that slightly customizes `manylinux2014` to our use case
- a readme giving a highlevel overview of what we are doing

### CODE REVIEW

File: contrib/packaging_python/Readme.md [hunk 1]
```diff
@@ -0,0 +1,123 @@
...
+
+You can remove the repository argument from the upload command to upload to
+real PyPi.
+
+## What goes on within the Docker container
+
+### Setup
+
+Compared to `manylinux2014`, the container comes with `ccache` and some cvc5
+dependencies preinstalled. Additionally, we point `cmake` to `ccache` using the `CMAKE_<LANG>_COMPILER_LAUNCHER` environment variables.
```
Comment (4tXJ7f) [inline contrib/packaging_python/Readme.md:62]: > Additionally, we point `cmake` to `ccache` using the `CMAKE_<LANG>_COMPILER_LAUNCHER` environment variables.

It looks like that is not accurate anymore. Instead, the container creates symlinks.

File: contrib/packaging_python/Readme.md [hunk 2]
```diff
@@ -0,0 +1,123 @@
...
+rm -r build
+# start virtualenv
+source ./env<Python version>/bin/activate
+# install the wheel
+pip install <path to wheel for this Python version>
+# run a small example to make sure it works
+python3 -c "import pycvc5; solver = pycvc5.Solver(); print(solver.getIntegerSort())"
+```
+
+### Test PyPi (in VM)
```
Comment (4tXJ7f) [inline contrib/packaging_python/Readme.md:88]:
```suggestion
### Test PyPi (in Docker)
```

File: contrib/packaging_python/Readme.md [hunk 3]
```diff
@@ -0,0 +1,123 @@
...
+Note that you will need a TestPyPi login. Once it has been uploaded, you can
+test (from anywhere, not just the container) that the wheel works by installing
+it from the TestPyPi repository.
+
+```
+python3 -m pip install --index-url https://test.pypi.org/simple/ pycvc5==<cvc5 version number>
+python3 -c "import pycvc5; solver = pycvc5.Solver(); print(solver.getIntegerSort())"
+```
+
+### Upload to PyPi (from VM)
```
Comment (4tXJ7f) [inline contrib/packaging_python/Readme.md:109]:
```suggestion
### Upload to PyPi (from Docker)
```

File: contrib/packaging_python/Readme.md [hunk 4]
```diff
@@ -0,0 +1,123 @@
...
+2. configures cvc5 appropriately using the `mk_build_dir.py` script, which makes sure that the venv python version is used;
+3. builds cvc5 and the python extension via the `mk_wheel.py` script;
+4. postprocesses the wheel (with `auditwheel`) and moves it out of the build folder.
+
+
+## Uploading to PyPi
+
+To upload a wheel to test PyPi,
+
+    twine upload --repository testpypi -u $USERNAME -p $PASSWORD PATH_TO_WHEEL
```
Comment (4tXJ7f) [inline contrib/packaging_python/Readme.md:43]: Nit:

```suggestion
    twine upload --repository testpypi -u $USERNAME -p $PASSWORD <path to wheel>
```

(for consistency with the other instructions)

File: contrib/packaging_python/mk_clean_wheel.sh [hunk 5]
```diff
@@ -0,0 +1,71 @@
...
+CONFIG="$2"
+PYVERSION=$($PYTHONBIN -c "import sys; print(sys.implementation.name + sys.version.split()[0])")
+
+# setup and activate venv
+echo "Making venv with $PYTHONBIN"
+ENVDIR=env$PYVERSION
+$PYTHONBIN -m venv ./$ENVDIR
+source ./$ENVDIR/bin/activate
+
+which python
```
Comment (4tXJ7f) [inline contrib/packaging_python/mk_clean_wheel.sh:36]: Is this intentional or some leftover from a debugging session?

**Review by 4tXJ7f — CHANGES_REQUESTED** [review]
Just some minor requests.

File: contrib/packaging_python/Readme.md (hunk 2 above)
Comment (nafur) [reply contrib/packaging_python/Readme.md:89]: We don't actually do this from within the container but from the CI...

File: contrib/packaging_python/mk_wheel.py [hunk 6]
```diff
@@ -0,0 +1,118 @@
...
+from setuptools.command.build_ext import build_ext
+from distutils.version import LooseVersion
+
+
+def get_project_src_path():
+    # expecting this script to be in contrib/packaging_python/
+    # The project source directory is two directories up
+    name = __file__
+    for i in range(3):
+        name = os.path.dirname(name)
```
Comment (mpreiner) [inline contrib/packaging_python/mk_wheel.py:41]: Can we maybe add checks to ensure that the directory returned is indeed the project source directory? E.g. check if the directory returned contains directory `src` or something like that?

File: contrib/packaging_python/mk_wheel.py [hunk 7]
```diff
@@ -0,0 +1,118 @@
...
+            self.build_extension(ext)
+
+    @staticmethod
+    def is_windows():
+        tag = platform.system().lower()
+        return tag == "windows"
+
+    def build_extension(self, ext):
+        # build the python api
+        subprocess.check_call(['cmake', '--build', '.', '--target', 'cvc5_python_api', '-j', '10'])
```
Comment (mpreiner) [inline contrib/packaging_python/mk_wheel.py:95]: Don't use a fixed number of threads for `-j`.

File: contrib/packaging_python/Readme.md [hunk 8]
```diff
@@ -0,0 +1,122 @@
...
+pip install <path to wheel for this Python version>
+# run a small example to make sure it works
+python3 -c "import pycvc5; solver = pycvc5.Solver(); print(solver.getIntegerSort())"
+```
+
+### Test PyPi
+
+In addition to the local test described above, you can do a test upload to
+[TestPyPi](https://packaging.python.org/guides/using-testpypi/). To do this from
+run the following after building the wheel:
```
Comment (4tXJ7f) [inline contrib/packaging_python/Readme.md:93]: Sentence is broken now.

File: contrib/packaging_python/mk_wheel.py (hunk 6 above)
Comment (4tXJ7f) [reply contrib/packaging_python/mk_wheel.py:41]: Hmm, we also have a `src` directory in our build folders. Maybe checking for `configure.sh` would be a bit safer?

File: contrib/packaging_python/mk_wheel.py [hunk 9]
```diff
@@ -40,7 +40,7 @@ def get_project_src_path():
     for i in range(3):
         name = os.path.dirname(name)
     name = os.path.abspath(name)
-    if not os.path.isdir(os.path.join(name, 'src')):
+    if not os.path.isdir(os.path.join(name, 'configure.sh')):
```
Comment (4tXJ7f) [inline contrib/packaging_python/mk_wheel.py:43]:
```suggestion
    if not os.path.isfile(os.path.join(name, 'configure.sh')):
```

;)

**Review by 4tXJ7f — APPROVED** [review]
Passt.

**Pull request merged**